common/chunkStatusLogger.go

```diff
@@ -333,16 +341,21 @@ func (csl *chunkStatusLogger) main(chunkLogPath string) {
 
 	alwaysFlushFromNowOn := false
 	for x := range csl.unsavedEntries {
-		if x == nil {
-			alwaysFlushFromNowOn = true
-			doFlush()
-			csl.flushDone <- struct{}{}
-			continue // TODO can become break (or be moved to later if we close unsaved entries, once we figure out how we got stuff written to us after CloseLog was called)
-		}
-		_, _ = w.WriteString(fmt.Sprintf("%s,%d,%s,%s\n", x.Name, x.OffsetInFile(), x.reason, x.waitStart))
-		if alwaysFlushFromNowOn {
-			// TODO: remove when we figure out how we got stuff written to us after CloseLog was called. For now, this should handle those cases (if they still exist)
-			doFlush()
+		select {
+		case <-csl.done:
+			return
+		default:
```

Comment:
Instead of doing the select here, we should do it inside the if x == nil block, so that if there are genuine chunks to be logged we log them before returning. Note that due to scheduling delays, we might have one or more chunks in the q followed by nil check, but the done channel may already have been signalled. We should only look at done or flushDone when x==nil.

Reply:
Changed to first flushlog , then we do the closing of goroutine.